core/server/src/main/java/alluxio/master/block/BlockMaster.java

```diff
@@ -316,10 +316,9 @@ public void removeBlocks(List<Long> blockIds, boolean delete) {
             continue;
           }
           for (long workerId : new ArrayList<Long>(masterBlockInfo.getWorkers())) {
-            masterBlockInfo.removeWorker(workerId);
```

Comment:
I think metadata maintenance should be based-on the feedback of workers' heartbeats. e.g. Master removes a worker from its BlockInfo only after receiving the acknowledgement from the corresponding heartbeat. Ideally, all remove/free operations should be catched at:
https://github.com/Alluxio/alluxio/blob/master/core/server/src/main/java/alluxio/master/block/BlockMaster.java#L619